File: src/input/pcapthread.c
/*
 * Copyright (c) 2018, OARC, Inc.
 * All rights reserved.
 *
 * This file is part of dnsjit.
 *
 * dnsjit is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * dnsjit is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with dnsjit.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include "input/pcapthread.h"
#include "core/tracking.h"
#include "core/object/packet.h"

#include <string.h>

static core_log_t         _log      = LOG_T_INIT("input.pcapthread");
static input_pcapthread_t _defaults = {
    LOG_T_INIT_OBJ("input.pcapthread"),
    0, 0,
    0, 0, 0,
    { 0, 0 }, { 0, 0 },
    0, 0, 0, 0,
    PCAP_THREAD_OK,
    0, 1
};

core_log_t* input_pcapthread_log()
{
    return &_log;
}

static void _udp(u_char* user, const pcap_thread_packet_t* packet, const u_char* payload, size_t length)
{
    input_pcapthread_t*  self = (input_pcapthread_t*)user;
    core_object_packet_t pkt  = CORE_OBJECT_PACKET_INIT(0);

    self->pkts++;

    if (self->only_queries) {
        if (length < 3) {
            self->drop++;
            return;
        }
        if (payload[2] & 0x80) {
            self->ignore++;
            return;
        }

        self->queries++;
    }

    if (packet->have_iphdr) {
        pkt.src_addr = &packet->iphdr.ip_src;
        pkt.dst_addr = &packet->iphdr.ip_dst;
        pkt.is_ipv6  = 0;
    } else if (packet->have_ip6hdr) {
        pkt.src_addr = &packet->ip6hdr.ip6_src;
        pkt.dst_addr = &packet->ip6hdr.ip6_dst;
        pkt.is_ipv6  = 1;
    }

    for (; packet; packet = packet->prevpkt) {
        if (packet->have_pkthdr) {
            pkt.ts.sec  = packet->pkthdr.ts.tv_sec;
            pkt.ts.nsec = packet->pkthdr.ts.tv_usec * 1000;
            break;
        }
    }
    if (!packet) {
        self->drop++;
        return;
    }

    pkt.src_id = self->src_id;
    if (!self->qr_id) {
        /* 0 is error */
        self->qr_id++;
    }
    pkt.qr_id = self->qr_id++;

    pkt.is_udp  = 1;
    pkt.is_tcp  = 0;
    pkt.sport   = packet->udphdr.uh_sport;
    pkt.dport   = packet->udphdr.uh_dport;
    pkt.payload = payload;
    pkt.len     = length;

    self->recv(self->ctx, (core_object_t*)&pkt);
}

static void _tcp(u_char* user, const pcap_thread_packet_t* packet, const u_char* payload, size_t length)
{
    input_pcapthread_t*  self = (input_pcapthread_t*)user;
    core_object_packet_t pkt  = CORE_OBJECT_PACKET_INIT(0);

    self->pkts++;

    // TODO:
    // - include DNS length
    // - track queries if length comes in its own packet
    // - keep length in raw if sent in same packet
    if (length < 3) {
        self->drop++;
        return;
    }
    payload += 2;
    length -= 2;

    if (self->only_queries) {
        if (length < 3) {
            self->drop++;
            return;
        }
        if (payload[2] & 0x80) {
            self->ignore++;
            return;
        }

        self->queries++;
    }

    if (packet->have_iphdr) {
        pkt.src_addr = &packet->iphdr.ip_src;
        pkt.dst_addr = &packet->iphdr.ip_dst;
        pkt.is_ipv6  = 0;
    } else if (packet->have_ip6hdr) {
        pkt.src_addr = &packet->ip6hdr.ip6_src;
        pkt.dst_addr = &packet->ip6hdr.ip6_dst;
        pkt.is_ipv6  = 1;
    }

    for (; packet; packet = packet->prevpkt) {
        if (packet->have_pkthdr) {
            pkt.ts.sec  = packet->pkthdr.ts.tv_sec;
            pkt.ts.nsec = packet->pkthdr.ts.tv_usec * 1000;
            break;
        }
    }
    if (!packet) {
        self->drop++;
        return;
    }

    pkt.src_id = self->src_id;
    if (!self->qr_id) {
        /* 0 is error */
        self->qr_id++;
    }
    pkt.qr_id = self->qr_id++;

    pkt.is_udp  = 0;
    pkt.is_tcp  = 1;
    pkt.sport   = packet->udphdr.uh_sport;
    pkt.dport   = packet->udphdr.uh_dport;
    pkt.payload = payload;
    pkt.len     = length;

    self->recv(self->ctx, (core_object_t*)&pkt);
}

int input_pcapthread_init(input_pcapthread_t* self)
{
    if (!self) {
        return 1;
    }

    *self        = _defaults;
    self->src_id = core_tracking_src_id();

    ldebug("init");

    if (!(self->pt = pcap_thread_create())) {
        return 1;
    }

    if ((self->err = pcap_thread_set_use_threads(self->pt, 1))
        || (self->err = pcap_thread_set_use_layers(self->pt, 1))
        || (self->err = pcap_thread_set_callback_udp(self->pt, _udp))
        || (self->err = pcap_thread_set_callback_tcp(self->pt, _tcp))
        || (self->err = pcap_thread_set_snaplen(self->pt, 64 * 1024))
        || (self->err = pcap_thread_set_buffer_size(self->pt, 4 * 1024 * 1024))
        || (self->err = pcap_thread_set_queue_mode(self->pt, PCAP_THREAD_QUEUE_MODE_DIRECT))) {
        return 1;
    }
    self->setup_ok = 1;

    return 0;
}

int input_pcapthread_destroy(input_pcapthread_t* self)
{
    if (!self) {
        return 1;
    }

    ldebug("destroy");

    pcap_thread_free(self->pt);

    return 0;
}

int input_pcapthread_open(input_pcapthread_t* self, const char* device)
{
    if (!self || !self->setup_ok) {
        return 1;
    }

    ldebug("open %s", device);

    if ((self->err = pcap_thread_open(self->pt, device, (void*)self)) != PCAP_THREAD_OK) {
        return 1;
    }

    return 0;
}

int input_pcapthread_open_offline(input_pcapthread_t* self, const char* file)
{
    if (!self || !self->setup_ok || !file) {
        return 1;
    }

    ldebug("open_offline %s", file);

    if ((self->err = pcap_thread_open_offline(self->pt, file, (void*)self)) != PCAP_THREAD_OK) {
        return 1;
    }

    return 0;
}

int input_pcapthread_run(input_pcapthread_t* self)
{
    struct timespec ts, te;
    if (!self || !self->setup_ok || !self->recv) {
        return 1;
    }

    ldebug("run");

    clock_gettime(CLOCK_MONOTONIC, &ts);
    if ((self->err = pcap_thread_run(self->pt)) != PCAP_THREAD_OK) {
        return 1;
    }
    clock_gettime(CLOCK_MONOTONIC, &te);

    self->ts.sec  = ts.tv_sec;
    self->ts.nsec = ts.tv_nsec;
    self->te.sec  = te.tv_sec;
    self->te.nsec = te.tv_nsec;

    return 0;
}

int input_pcapthread_next(input_pcapthread_t* self)
{
    if (!self || !self->setup_ok || !self->recv) {
        return 1;
    }

    ldebug("next");

    if ((self->err = pcap_thread_next(self->pt)) != PCAP_THREAD_OK) {
        return 1;
    }

    return 0;
}

const char* input_pcapthread_errbuf(input_pcapthread_t* self)
{
    if (!self || !self->setup_ok) {
        return 0;
    }

    return pcap_thread_errbuf(self->pt);
}

const char* input_pcapthread_strerr(int err)
{
    return pcap_thread_strerr(err);
}

File: src/input/pcapthread.lua
-- Copyright (c) 2018, OARC, Inc.
-- All rights reserved.
--
-- This file is part of dnsjit.
--
-- dnsjit is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
--
-- dnsjit is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with dnsjit.  If not, see <http://www.gnu.org/licenses/>.

-- dnsjit.input.pcapthread
-- Read input from an interface or PCAP file using pcap-thread
--   local input = require("dnsjit.input.pcapthread").new()
--   input:open_offline("file.pcap")
--   input:receiver(filter_or_output)
--   input:run()
--
-- Input module for reading packets from interfaces and PCAP files.
module(...,package.seeall)

require("dnsjit.input.pcapthread_h")
local ffi = require("ffi")
local C = ffi.C

local t_name = "input_pcapthread_t"
local input_pcapthread_t = ffi.typeof(t_name)
local Pcapthread = {}

-- Create a new Pcapthread input.
function Pcapthread.new()
    local self = {
        _receiver = nil,
        obj = input_pcapthread_t(),
    }
    C.input_pcapthread_init(self.obj)
    ffi.gc(self.obj, C.input_pcapthread_destroy)
    return setmetatable(self, { __index = Pcapthread })
end

-- Return the Log object to control logging of this instance or module.
function Pcapthread:log()
    if self == nil then
        return C.input_pcapthread_log()
    end
    return self.obj._log
end

-- Set the receiver to pass queries to.
function Pcapthread:receiver(o)
    self.obj._log:debug("receiver()")
    self.obj.recv, self.obj.ctx = o:receive()
    self._receiver = o
end

-- Enable (true) or disable (false) only passing DNS queries to receiver, if
-- .I bool
-- is not specified then return if only passing DNS queries is on (true) or
-- off (false).
function Pcapthread:only_queries(bool)
    if bool == nil then
        if self.obj.only_queries == 1 then
            return true
        end
        return false
    elseif bool == true then
        self.obj.only_queries = 1
    else
        self.obj.only_queries = 0
    end
end

-- Return the snaphot length, see
-- .BR pcap_snapshot (3pcap)
-- for more information.
function Pcapthread:snapshot()
    return C.pcap_thread_snapshot(self.obj.pt)
end

-- Set the number of bytes to capture and return 0 if successful, if
-- .I len
-- is not specified then return the number of bytes that was previously
-- set by this function.
-- Use
-- .BR snaphot ()
-- too see how many bytes are actually captured.
function Pcapthread:snaplen(len)
    if len == nil then
        return C.pcap_thread_snaplen(self.obj.pt)
    end
    return C.pcap_thread_set_snaplen(self.obj.pt, len)
end

-- Enable (true) or disable (false) promiscuous mode and return 0 if
-- successful, if
-- .I bool
-- is not specified then return if promiscuous mode is on (true) or
-- off (false).
-- See
-- .BR pcap (3pcap)
-- for more information.
function Pcapthread:promiscuous(bool)
    if bool == nil then
        if C.pcap_thread_promiscuous(self.obj.pt) == 1 then
            return true
        end
        return false
    elseif bool == true then
        return C.pcap_thread_set_promiscuous(self.obj.pt, 1)
    else
        return C.pcap_thread_set_promiscuous(self.obj.pt, 0)
    end
end

-- Enable (true) or disable (false) monitor mode and return 0 if
-- successful, if
-- .I bool
-- is not specified then return if monitor mode is on (true) or off (false).
-- See
-- .BR pcap (3pcap)
-- for more information.
function Pcapthread:monitor(bool)
    if bool == nil then
        if C.pcap_thread_monitor(self.obj.pt) == 1 then
            return true
        end
        return false
    elseif bool == true then
        return C.pcap_thread_set_monitor(self.obj.pt, 1)
    else
        return C.pcap_thread_set_monitor(self.obj.pt, 0)
    end
end

-- Set the timeout in milliseconds and return 0 if successful, if
-- .I ms
-- is not specified then return the current timeout.
-- See
-- .BR pcap_set_timeout (3pcap)
-- for more information.
function Pcapthread:timeout(ms)
    if ms == nil then
        return C.pcap_thread_timeout(self.obj.pt)
    end
    return C.pcap_thread_set_timeout(self.obj.pt, ms)
end

-- Set the buffer size and return 0 if successful, if
-- .I size
-- is not specified then return the current buffer size.
-- See
-- .BR pcap_set_buffer_size (3pcap)
-- for more information.
function Pcapthread:buffer_size(size)
    if size == nil then
        return C.pcap_thread_buffer_size(self.obj.pt)
    end
    return C.pcap_thread_set_buffer_size(self.obj.pt, size)
end

-- Enable (true) or disable (false) immediate mode and return 0 if
-- successful, if
-- .I bool
-- is not specified then return if immediate mode is on (true) or off (false).
-- May have no effect depending on libpcap version.
-- See
-- .BR pcap_set_immediate_mode (3pcap)
-- for more information.
function Pcapthread:immediate_mode()
    if bool == nil then
        if C.pcap_thread_immediate_mode(self.obj.pt) == 1 then
            return true
        end
        return false
    elseif bool == true then
        return C.pcap_thread_set_immediate_mode(self.obj.pt, 1)
    else
        return C.pcap_thread_set_immediate_mode(self.obj.pt, 0)
    end
end

-- Set the PCAP packet filter to use and return 0 if successful, if
-- .I pf
-- is not specified then return the current packet filter in use.
-- If
-- .I pf
-- is false then clear the packet filter and return 0 on success.
-- See
-- .BR pcap-filter (3pcap)
-- for more information.
function Pcapthread:filter(pf)
    if pf == nil then
        return ffi.string(C.pcap_thread_filter(self.obj.pt))
    elseif pf == false then
        return C.pcap_thread_clear_filter(self.obj.pt)
    end
    return C.pcap_thread_set_filter(self.obj.pt, pf, string.len(pf))
end

-- Return the error number return from libpcap while parsing the packet
-- filter.
function Pcapthread:filter_errno()
    return C.pcap_thread_filter_errno(self.obj.pt)
end

-- Enable (true) or disable (false) packet filter optimizing and return 0
-- if successful, if
-- .I bool
-- is not specified then return if packet filter optimizing is on (true) or
-- off (false).
-- See
-- .BR pcap_compile (3pcap)
-- for more information.
function Pcapthread:filter_optimize(bool)
    if bool == nil then
        if C.pcap_thread_filter_optimize(self.obj.pt) == 1 then
            return true
        end
        return false
    elseif bool == true then
        return C.pcap_thread_set_filter_optimize(self.obj.pt, 1)
    else
        return C.pcap_thread_set_filter_optimize(self.obj.pt, 0)
    end
end

-- Set the network mask to give when compiling the packet filter and
-- return 0 if successful, if
-- .I netmask
-- is not specified then return the current network mask.
-- See
-- .BR pcap_compile (3pcap)
-- for more information.
function Pcapthread:filter_netmask(netmask)
    if netmask == nil then
        return C.pcap_thread_filter_netmask(self.obj.pt)
    end
    return C.pcap_thread_set_filter_netmask(self.obj.pt, netmask)
end

-- Open an interface device for capturing, can be given multiple times to
-- open additional interfaces.
-- Returns 0 on success.
function Pcapthread:open(device)
    return C.input_pcapthread_open(self.obj, device)
end

-- Open a PCAP file for processing, can be given multiple times to
-- open additional files.
-- Returns 0 on success.
function Pcapthread:open_offline(file)
    return C.input_pcapthread_open_offline(self.obj, file)
end

-- Start processing packets from opened devices and PCAP files.
-- Returns 0 on success.
function Pcapthread:run()
    return C.input_pcapthread_run(self.obj)
end

-- Process one packet from the opened devices and PCAP files, the opened
-- sources are processed as a round robin list.
-- Returns 0 on success.
function Pcapthread:next()
    return C.input_pcapthread_next(self.obj)
end

-- Return the last error as a string that came from libpcap functions
-- that takes an error buffer.
-- See for example
-- .BR pcap_open_offline (3pcap).
function Pcapthread:errbuf()
    return ffi.string(C.input_pcapthread_errbuf(self.obj))
end

-- Return the error
-- .I err
-- as a string or if not specified the last error.
function Pcapthread:strerr(err)
    if err == nil then
        return ffi.string(C.input_pcapthread_strerr(self.obj.err))
    end
    return ffi.string(C.input_pcapthread_strerr(err))
end

-- Return the seconds and nanoseconds (as a list) of the start time for
-- .BR Pcapthread:run() .
function Pcapthread:start_time()
    return tonumber(self.obj.ts.sec), tonumber(self.obj.ts.nsec)
end

-- Return the seconds and nanoseconds (as a list) of the stop time for
-- .BR Pcapthread:run() .
function Pcapthread:end_time()
    return tonumber(self.obj.te.sec), tonumber(self.obj.te.nsec)
end

-- Return the number of packets seen.
function Pcapthread:packets()
    return tonumber(self.obj.pkts)
end

-- Return the number of packets dropped.
function Pcapthread:dropped()
    return tonumber(self.obj.drop)
end

-- Return the number of packets ignored as a result of only processing queries.
-- See
-- .BR Pcapthread:only_queries() .
function Pcapthread:ignored()
    return tonumber(self.obj.ignore)
end

-- Return the number of queries seen, see
-- .BR Pcapthread:only_queries() .
function Pcapthread:queries()
    return tonumber(self.obj.queries)
end

return Pcapthread
